### PR TITLE
BigInt writing support, and support reading from buffer offset

### PR DESCRIPTION
**Pull request by extremeheat**
BigInt writing support, and support reading from buffer offset

If you're reading from a stream of data, it's easier and more efficient to increment offset position than make a new buffer slice every time.

### CODE REVIEW

**Comment by rom1504** [comment]
linting not passing

File: src/compiler.js [hunk 1]
```diff
@@ -95,8 +95,8 @@ class CompiledProtodef {
     return buffer
   }
 
-  parsePacketBuffer (type, buffer) {
-    const { value, size } = tryCatch(() => this.read(buffer, 0, type),
+  parsePacketBuffer (type, buffer, offset = 0) {
```
Comment (rom1504) [inline src/compiler.js:98]: can you update the doc for this change in API ?

File: src/datatypes/numeric.js [hunk 2]
```diff
@@ -9,8 +9,12 @@ function readI64 (buffer, offset) {
 }
 
 function writeI64 (value, buffer, offset) {
-  buffer.writeInt32BE(value[0], offset)
-  buffer.writeInt32BE(value[1], offset + 4)
+  if (typeof value === 'bigint') {
```
Comment (rom1504) [inline src/datatypes/numeric.js:12]: what about reading?
This introduced an inconsistency between values read by i64 and the ones written by it

**Review by rom1504 — CHANGES_REQUESTED** [review]

File: src/compiler.js (hunk 1 above)
Comment (extremeheat) [reply src/compiler.js:98]: done

File: src/datatypes/numeric.js (hunk 2 above)
Comment (extremeheat) [reply src/datatypes/numeric.js]: Reading would be a breaking change, would also need to bump the node version to 10.4

Comment (rom1504) [reply src/datatypes/numeric.js]: yes. I'm not sure it makes sense to change writer but not reader
How do you then use that?

Comment (extremeheat) [reply src/datatypes/numeric.js]: In PrismarineJS/prismarine-nbt#35 I have it reading 64-bit zigzag varints into a BigInt to keep things simple and prevent an overflow. So to make sure that the file can be written back in a non-varint format I added support for writing them here, but yeah it looks like it could cause API issues. Just thought it would be a bit wasteful to have to do extra conversions between number/bigint in the internal varint implementation

Comment (extremeheat) [reply src/datatypes/numeric.js]: okay so I've been looking at converting to arrays in pnbt but after thinking about it for a bit I can't really think of any good use case to send back arrays. If the user has to convert them to bigint to do anything useful with them (e.g. to read/update a timestamp) and then the user has to send back an array to pnbt, and then that has to be converted into a bigint, then that seems a bit excessive , especially as operations on big numbers in JS is already slow.

File: src/datatypes/numeric.js [hunk 3]
```diff
@@ -9,8 +9,12 @@ function readI64 (buffer, offset) {
...
 
 function writeI64 (value, buffer, offset) {
-  buffer.writeInt32BE(value[0], offset)
-  buffer.writeInt32BE(value[1], offset + 4)
+  if (typeof value === 'bigint') { // eslint-disable-line
+    buffer.writeBigInt64BE(value, offset)
+  } else {
+    buffer.writeInt32BE(value[0], offset)
+    buffer.writeInt32BE(value[1], offset + 4)
+  }
```
Comment (Saiv46) [inline src/datatypes/numeric.js:17]: This function is redurant

File: src/datatypes/numeric.js [hunk 4]
```diff
@@ -23,8 +27,12 @@ function readLI64 (buffer, offset) {
...
 
 function writeLI64 (value, buffer, offset) {
-  buffer.writeInt32LE(value[0], offset + 4)
-  buffer.writeInt32LE(value[1], offset)
+  if (typeof value === 'bigint') { // eslint-disable-line
+    buffer.writeBigInt64LE(value, offset)
+  } else {
+    buffer.writeInt32LE(value[0], offset + 4)
+    buffer.writeInt32LE(value[1], offset)
+  }
```
Comment (Saiv46) [inline src/datatypes/numeric.js:35]: This function is redurant

File: src/datatypes/numeric.js [hunk 5]
```diff
@@ -37,8 +45,12 @@ function readU64 (buffer, offset) {
...
 
 function writeU64 (value, buffer, offset) {
-  buffer.writeUInt32BE(value[0], offset)
-  buffer.writeUInt32BE(value[1], offset + 4)
+  if (typeof value === 'bigint') { // eslint-disable-line
+    buffer.writeBigUInt64BE(value, offset)
+  } else {
+    buffer.writeUInt32BE(value[0], offset)
+    buffer.writeUInt32BE(value[1], offset + 4)
+  }
```
Comment (Saiv46) [inline src/datatypes/numeric.js:53]: This function is redurant

File: src/datatypes/numeric.js [hunk 6]
```diff
@@ -51,8 +63,12 @@ function readLU64 (buffer, offset) {
...
 
 function writeLU64 (value, buffer, offset) {
-  buffer.writeUInt32LE(value[0], offset + 4)
-  buffer.writeUInt32LE(value[1], offset)
+  if (typeof value === 'bigint') { // eslint-disable-line
+    buffer.writeBigUInt64LE(value, offset)
+  } else {
+    buffer.writeUInt32LE(value[0], offset + 4)
+    buffer.writeUInt32LE(value[1], offset)
+  }
```
Comment (Saiv46) [inline src/datatypes/numeric.js:71]: This function is redurant

File: src/datatypes/numeric.js [hunk 7]
```diff
@@ -66,6 +82,7 @@ function generateFunctions (bufferReader, bufferWriter, size, schema) {
     }
   }
   const writer = (value, buffer, offset) => {
+    if (typeof value === 'bigint') value = Number(value) // eslint-disable-line
```
Comment (Saiv46) [inline src/datatypes/numeric.js:85]: Bug ahead - accuracy after `Number.MAX_SAFE_INTEGER` will be lost

**Review by Saiv46 — CHANGES_REQUESTED** [review]
This is a breaking change. Can I help you finish this PR?

File: src/datatypes/numeric.js (hunk 7 above)
Comment (extremeheat) [reply src/datatypes/numeric.js:85]: 64-bit integers wouldn't be written here, so there's no issue of overflow

**Comment by extremeheat** [comment]
Sure, I'm not aware of any breaking changes here, the existing logic remains the same with the ability to serialize BigInt with the native support in Buffer. If BigInt isn't supported (node <10) the typeof matches should fail. I'll look at this a bit more in the weekend, but I currently depend on some of this functionality for some of my projects through pnbt, so the options to handle read/write 64-bit integers comes down to 1) keep arrays and ignore the upper 32 bits or 2) do conversion back and forth.

File: src/datatypes/numeric.js (hunk 7 above)
Comment (Saiv46) [reply src/datatypes/numeric.js:85]: Why would you leave functions, when you can add the following lines to `nums`:
```
  'i64': ['readBigInt64BE', 'writeBigInt64BE', 8],
  'u64': ['readBigUInt64BE', 'writeBigUInt64BE', 8],
  'li64': ['readBigInt64LE', 'writeBigInt64LE', 8],
  'lu64': ['readBigUInt64LE', 'writeBigUInt64LE', 8],
```

Comment (rom1504) [reply src/datatypes/numeric.js:85]: That would be a breaking change

Comment (Saiv46) [reply src/datatypes/numeric.js:85]: Isn't that already a breaking change?

Comment (Saiv46) [reply src/datatypes/numeric.js:85]: @extremeheat Okay, I found a way how to keep two behaviors at same time
```js
class BigIntExtended {
    constructor (arg, isLE) {
        this.bint = BigInt(arg)
        if (isLE) {
            this[0] = this.upper
            this[1] = this.lower
        } else {
            this[0] = this.lower
            this[1] = this.upper
        }
    }
    get lower () { return Number(this.bint & 0xFFFFFFFFn) }
    get upper () { return Number((this.bint >> 32n) & 0xFFFFFFFFn) }
    valueOf () { return this.bint }
}
```

Comment (extremeheat) [reply src/datatypes/numeric.js:85]: Yeah, this seems interesting. So in order to write, the check would be typeof var.valueOf() instead, to read it would read  as a BigInt and return this class. Also, a toJSON function could be added to make it serializable. Would be particularly useful for https://github.com/PrismarineJS/mineflayer/pull/1543 I think. cc opinion of @rom1504 @Karang

Comment (extremeheat) [reply src/datatypes/numeric.js:85]: @Saiv46 you updated it, why do you do the shifting in valueOf? arg can just be stored and returned by valueOf(), and array access would still work. What do you think about just

```js
class BigIntExtended extends Array {
    constructor (arg, isLE) {
        const arr = [Number((arg >> 32n) & 0xFFFFFFFFn), Number(arg & 0xFFFFFFFFn)]
        if (isLE) arr.reverse()
        super(...arr)
        this.arg = arg
        this.isLE = isLE | 0
        this.valueOf = () => arg
    }
}
```
?

**Comment by extremeheat** [comment]
Tried doing using the new BigIntExt method from @Saiv46 , but this require node.js engine needs to be updated to 10.4. This still causes standardjs to fail because the version of standardjs the project is using (12) is incompatible with node 10.4. Updating standard to the latest version would change alot of the code, and should probably be done outside this PR, so that's the major blocker here.

**Comment by extremeheat** [comment]
Updated the PR to use a fixed version of BigIntExt, looks like protodef expects signed integers. Also the isLE check is not needed as 
 it's always big endian when shifting, it would only be needed to read from the stream. Until the standardjs issue gets fixed I added it to the lint ignore list so the tests pass.

**Comment by rom1504** [comment]
can you please explain what you did with bigint ?
This looks breaking to me.

**Comment by extremeheat** [comment]
Fixed valueOf and simplified it a bit further. I don't think there should be any breaking changes, though it did look more complicated than it needed to.

Basically it returns an array that has a primitive value of a BigInt so when you do <>+-=/%, etc it acts as a BigInt, but it's still an array so array accessing/modifications/serialization still work

```js
class BigIntExtended extends Array {
  static fromInt(arg) { // just so this can be created from a number/bigint
    if (typeof arg === 'number') arg = BigInt(arg)
    // converts bigint into array:
    return new BigIntExtended(Number(BigInt.asIntN(32, arg >> 32n)), Number(BigInt.asIntN(32, arg)))
  }
  // creates a bigint from array:
  valueOf() { return BigInt.asIntN(64, BigInt(this[0]) << 32n) | BigInt.asUintN(32, BigInt(this[1])) }
}

// usage

var be = BigIntExtended.fromInt(-22)
console.log(be + 64n) // ✔ 42n
console.log(be, be[0], be[1]) // BigIntExtended(2) [-1, -22]  -1  -22
console.log(be > -30n) // true
be[1] += 64 // lower half becomes positive
console.log(be +0n) // -4294967254n (coerced to a bigint)
console.log(be > -30n) // false
be[0] += 1
JSON.stringify(be) // "[0,42]"
be.toString() // "0,42"
be.valueOf() // 42n
```

File: src/datatypes/numeric.js [hunk 8]
```diff
@@ -1,58 +1,82 @@
...
-    value: [buffer.readInt32BE(offset), buffer.readInt32BE(offset + 4)],
+    value: new BigIntExtended(buffer.readInt32BE(offset), buffer.readInt32BE(offset + 4)),
     size: 8
   }
 }
 
 function writeI64 (value, buffer, offset) {
-  buffer.writeInt32BE(value[0], offset)
-  buffer.writeInt32BE(value[1], offset + 4)
+  if (typeof value.valueOf() === 'bigint') {
```
Comment (rom1504) [inline src/datatypes/numeric.js:17]: This way to check if it's a bigint seems costly.
It's doing the conversion to bigint and not just checking the type
Can't this just check instanceof BigIntExtended or BigInt instead ?

**Comment by rom1504** [comment]
It seems the ci is not running, I'll fix it after this PR.
Can you run npm test ?
And can you run npm run benchmark before/after this PR ?

**Comment by extremeheat** [comment]
All the tests are passing,

Benchmark on big:

```
> mocha benchmark/benchmark_unified.js



read/write x 1,937 ops/sec ±18.61% (70 runs sampled)
  √ read/write (5698ms)
read/write (compiled) x 11,513 ops/sec ±5.66% (72 runs sampled)
  √ read/write (compiled) (5690ms)

  2 passing (11s)
```

on master:

```
> mocha benchmark/benchmark_unified.js



read/write x 2,085 ops/sec ±14.73% (72 runs sampled)
  √ read/write (5640ms)
read/write (compiled) x 11,117 ops/sec ±13.91% (70 runs sampled)
  √ read/write (compiled) (5528ms)

  2 passing (11s)


```

**gitpod**
big:
```
> protodef@1.8.3 benchmark /workspace/node-protodef
> mocha benchmark/benchmark_unified.js



read/write x 5,982 ops/sec ±5.10% (78 runs sampled)
  ✓ read/write (5717ms)
read/write (compiled) x 25,757 ops/sec ±2.30% (86 runs sampled)
  ✓ read/write (compiled) (5736ms)

  2 passing (11s)
```

master:
```
> protodef@1.8.3 benchmark /workspace/node-protodef
> mocha benchmark/benchmark_unified.js



read/write x 6,555 ops/sec ±3.62% (80 runs sampled)
  ✓ read/write (5730ms)
read/write (compiled) x 26,793 ops/sec ±2.65% (88 runs sampled)
  ✓ read/write (compiled) (5802ms)

  2 passing (12s)
```

File: src/datatypes/numeric.js (hunk 8 above)
Comment (extremeheat) [reply src/datatypes/numeric.js]: ah yeah, the shifting could be an issue, updated to instanceof

File: src/datatypes/numeric.js [hunk 9]
```diff
@@ -1,58 +1,82 @@
 const { PartialReadError } = require('../utils')
 
+class BigIntExtended extends Array {
+  valueOf () { return BigInt.asIntN(64, BigInt(this[0]) << 32n) | BigInt.asUintN(32, BigInt(this[1])) }
+}
+
 function readI64 (buffer, offset) {
   if (offset + 8 > buffer.length) { throw new PartialReadError() }
+  const val = buffer.readBigInt64BE(offset)
```
Comment (rom1504) [inline src/datatypes/numeric.js:9]: This line seems unnecessary

File: src/datatypes/numeric.js [hunk 10]
```diff
@@ -1,58 +1,82 @@
...
+  } else {
+    buffer.writeInt32BE(value[0], offset)
+    buffer.writeInt32BE(value[1], offset + 4)
+  }
   return offset + 8
 }
 
 function readLI64 (buffer, offset) {
   if (offset + 8 > buffer.length) { throw new PartialReadError() }
+  const val = buffer.readBigInt64LE(offset)
```
Comment (rom1504) [inline src/datatypes/numeric.js:28]: Unused variable

File: src/datatypes/numeric.js [hunk 11]
```diff
@@ -1,58 +1,82 @@
...
+  } else {
+    buffer.writeInt32LE(value[0], offset + 4)
+    buffer.writeInt32LE(value[1], offset)
+  }
   return offset + 8
 }
 
 function readU64 (buffer, offset) {
   if (offset + 8 > buffer.length) { throw new PartialReadError() }
+  const val = buffer.readBigUInt64BE(offset)
```
Comment (rom1504) [inline src/datatypes/numeric.js:47]: Unused

File: src/datatypes/numeric.js [hunk 12]
```diff
@@ -1,58 +1,82 @@
...
+  } else {
+    buffer.writeUInt32BE(value[0], offset)
+    buffer.writeUInt32BE(value[1], offset + 4)
+  }
   return offset + 8
 }
 
 function readLU64 (buffer, offset) {
   if (offset + 8 > buffer.length) { throw new PartialReadError() }
+  const val = buffer.readBigUInt64LE(offset)
```
Comment (rom1504) [inline src/datatypes/numeric.js:66]: Unused

File: src/datatypes/numeric.js [hunk 13]
```diff
@@ -66,6 +90,7 @@ function generateFunctions (bufferReader, bufferWriter, size, schema) {
     }
   }
   const writer = (value, buffer, offset) => {
+    if (typeof value === 'bigint') value = Number(value)
```
Comment (rom1504) [inline src/datatypes/numeric.js:93]: Why is this needed ?
This is happening for all numeric types

**Comment by extremeheat** [comment]
removed the extra code

**Review by rom1504 — APPROVED** [review]
